Common/include/toolboxes/CQuasiNewtonDriver.hpp

```diff
@@ -0,0 +1,253 @@
+ *
+ * The SU2 Project is maintained by the SU2 Foundation
+ * (http://su2foundation.org)
+ *
+ * Copyright 2012-2020, SU2 Contributors (cf. AUTHORS.md)
+ *
+ * SU2 is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * SU2 is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with SU2. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "../omp_structure.hpp"
+#include "../mpi_structure.hpp"
+#include "CSymmetricMatrix.hpp"
+
+/*!
+ * \class A quasi-Newton fixed-point (FP) accelerator based on IQN-ILS.
+ * \note The implementation prioritizes storage, the LS problem is solved
+ * via the normal equations as that is easy to make parallel over MPI, it
+ * may however be unstable with large sample sizes (compared to QR decomp).
+ * Usage: Allocate, store the initial solution (operator (i,j), default is 0),
+ * run the FP, store its result ("FPresult"), compute new solution, use it
+ * as the new input of the FP, run the FP, etc.
+ */
+template<class Scalar_t>
+class CQuasiNewtonDriver {
+public:
+  using Scalar = Scalar_t;
+  using Index = typename su2matrix<Scalar>::Index;
+  static_assert(std::is_floating_point<Scalar>::value,"");
+
+private:
+  using MPI_Wrapper = typename SelectMPIWrapper<Scalar>::W;
+
+  enum: size_t {BLOCK_SIZE = 1024};     /*!< \brief Loop tiling parameter. */
+  std::vector<su2matrix<Scalar> > X, R; /*!< \brief Input and residual history of the FP. */
+  su2matrix<Scalar> corr;               /*!< \brief Correction to the natural FP result. */
+  su2vector<Scalar> mat, rhs, sol;      /*!< \brief Matrix, rhs, and solution of the normal equations. */
+  Index iSample = 0;                    /*!< \brief Current sample index. */
+  Index nPtDomain = 0;                  /*!< \brief Local size of the history, considered in dot products. */
+
+  void shiftHistoryLeft() {
+    for (Index i = 1; i < X.size(); ++i) {
+      /*--- Swap instead of moving to re-use the memory of the first sample.
+       * This is why X and R are not stored as contiguous blocks of mem. ---*/
+      std::swap(X[i-1], X[i]);
+      std::swap(R[i-1], R[i]);
+    }
+  }
+
+  void computeNormalEquations() {
+    /*--- Size for the dot products. ---*/
+    const auto end = std::min<Index>(nPtDomain,corr.rows())*corr.cols();
+
+    mat = Scalar(0); rhs = Scalar(0);
+
+    /*--- Tiled part of the loop. ---*/
+    Index begin = 0;
+    while (end-begin >= BLOCK_SIZE) {
+      computeNormalEquations<BLOCK_SIZE>(mat, rhs, begin);
+      begin += BLOCK_SIZE;
+    }
```

Comment:
For those interested in learning about performance techniques, I use something here called loop tiling, or loop blocking, or strip mining.
We need to compute A^T A where A is narrow, doing the natural "row dot column" algorithm is very inefficient because it makes poor use of the CPU cache, i.e. after doing the dot product of column 0 with itself, we go back to the beginning of column 0 to dot it with column 1, but by now the beginning of column 0 was evicted from cache (getting data from main memory is orders of magnitude slower than getting it from cache).
By tiling the loop we go over all the different combinations of columns and rows without evicting anything from the cache (since we work on much smaller sizes).
For general matrix multiplication we would need to tile both rows and columns.